Mirror disk dump / load helper

gavin-norman-sociomantic opened the pull request:
Rough concept. Open to all feedback as to the API / implementation / usefulness.

joseph-wakeling-sociomantic commented on src/dhtproto/client/helper/ChannelSerializer.d:
```diff
@@ -0,0 +1,354 @@
+
+    /// TODO
+    private alias void delegate ( void delegate ( ref hash_t, ref S ) )
+        DumpRecordDg;
```
Might be worth supporting `Contiguous!(S)` as dump input as well, since that may well be how the data is received and stored by the application.

gavin-norman-sociomantic replied:
Ah yes, good point. That ought to be the standard API.

joseph-wakeling-sociomantic commented on src/dhtproto/client/helper/ChannelSerializer.d:
```diff
@@ -0,0 +1,354 @@
+
+    ***************************************************************************/
+
+    public bool dump ( HashMap!(S) map, bool block = false )
```
Bear in mind that applications may be using other cache types than built-in AA's or `HashMap` (e.g. `StandardKeyKashingMap`).

gavin-norman-sociomantic replied:
Ah yes, I was thinking that `StandardKeyKashingMap` was dervied from `HashMap`, but it's not.

Arbitrary types are supported too, by the last overload.

joseph-wakeling-sociomantic replied:
Gotcha, makes sense.  Although it might be possible to simplify the different overloads somewhat by implementing a generic mechanism that supports any type that can be `foreach`'d over.

gavin-norman-sociomantic replied:
The map and AA overloads are purely convenience wrappers around the overloads for generic types.

joseph-wakeling-sociomantic replied:
That said in some ways I feel it might be simpler for the helper to avoid any need to care about the actual cache, and just provide the means to dump individual key/value pairs.

gavin-norman-sociomantic replied:
Yeah, the trouble is the dumping of an individual record has to occur inside `SafeFork`, so it's not something that can be called via the public API. The user has to register their intention to dump, then the `SafeFork` instance asks them for the records to be dumped, when it's ready.

gavin-norman-sociomantic commented:
@scott-gibson-sociomantic could you or someone from your team have a look at this? It's designed for your applications that need a complete mirror of a DHT channel at start up.

Commit:
Add ChannelSerializer helper

Designed for use by applications that require a complete copy of some DHT
channel before they are able to start operating.

Fixes #82.

gavin-norman-sociomantic commented:
Rebased onto v14.x.x. (v13.x.x is out of support.)

andrej-mitrovic-sociomantic commented on src/dhtproto/client/helper/ChannelSerializer.d:
```diff
@@ -0,0 +1,382 @@
+    {
+        void loadRecord ( hash_t k, ref Contiguous!(S) v )
+        {
+            aa[k] = v;
```
Is there any memory management you have to do here? For example, when loading keys/values which have arrays, would you have to manually duplicated them, or is this already handled somewhere else?

andrej-mitrovic-sociomantic replied:
In Alligator we used to have to do this manually, because we used the MapSerializer which uses a buffer for the keys / values, and there we had to copy the arrays manually.

gavin-norman-sociomantic replied:
I believe it's handled by the deserialization code below (https://github.com/sociomantic-tsunami/dhtproto/pull/162/commits/4a31c38d3e3af1f71b07b69741992af9c9394137#diff-5c38ebc93bed0b168f5f68fe43193a84R240), but we should definitely confirm that.

gavin-norman-sociomantic replied:
Yep. Looks like `v` (https://github.com/sociomantic-tsunami/dhtproto/pull/162/commits/4a31c38d3e3af1f71b07b69741992af9c9394137#diff-5c38ebc93bed0b168f5f68fe43193a84R236) will initially contain an empty buffer, which will be allocated and filled with the deserialized struct. The reference to this is then copied into the AA or map.

gavin-norman-sociomantic replied:
This is one of the nice things about working with `Contiguous!(S)`, by the way.

andrej-mitrovic-sociomantic replied:
Nice.

gavin-norman-sociomantic commented:
Finalised and unblocked.

gavin-norman-sociomantic commented:
As this PR implements a helper that is needed for some apps to support partial DHT handshake effectively, I feel we should merge it without application testing. When we come to adapt some app to use it, we can iron out any issues that come up.

matthias-wende-sociomantic commented on src/dhtproto/client/helper/ChannelSerializer.d:
```diff
@@ -0,0 +1,381 @@
+
+    private void forkEntryPoint ( )
+    {
+        this.current_record_dump_dg = this.new_record_dump_dg;
```
This is not needed due to the fork.

matthias-wende-sociomantic reviewed (commented):
Looks good. Just one comment.

gavin-norman-sociomantic commented:
I just noticed this PR doesn't implement the second feature mentioned in https://github.com/sociomantic-tsunami/dhtproto/issues/82#issuecomment-380724510.

>A means of handling errors while writing.

Example solutions:
1. Make a backup of the old dump before writing the new one, and load the backup if an error occurs while loading the dump.
2. Use the approach used in the dhtnode: write dumps to a temp file and rename it on success.

Commit:
Add test for channel serializer

scott-gibson-sociomantic commented:
I just realised, this PR uses no other swarm/dhtproto imports? Why does this live in dhtproto?

matthias-wende-sociomantic commented:
We talked about that and yes it is totally independent :). AFAIR the reason is that the use case of this helper is mainly restricted to the DHT. If someone comes up with a different use case independent of dhtproto it might be worth to consider moving it to ocean.

gavin-norman-sociomantic commented:
Exactly. The plan was/is also to enhance this to integrate it with the DHT client / channel mirror. (Though this module may remain independent.)
